## Joining Data with pandas




## Course Description

Being able to combine and work with multiple datasets is an essential skill for any aspiring Data Scientist. Pandas is a crucial cornerstone of the Python data science ecosystem, with Stack Overflow recording 5 million views for pandas questions. Learn to handle multiple DataFrames by combining, organizing, joining, and reshaping them using pandas. You'll work with datasets from the World Bank and the City Of Chicago. You will finish the course with a solid skillset for data-joining in pandas.

##  Data Merging Basics
Free
0%

Learn how you can merge disparate data using inner joins. By combining information from multiple sources you’ll uncover compelling insights that may have previously been hidden. You’ll also learn how the relationship between those sources, such as one-to-one or one-to-many, can affect your result.

    Inner join    50 xp
    What column to merge on?    50 xp
    Your first inner join    100 xp
    Inner joins and number of rows returned    100 xp
    One-to-many relationships    50 xp
    One-to-many classification    100 xp
    One-to-many merge    100 xp
    Merging multiple DataFrames    50 xp
    Total riders in a month    100 xp
    Three table merge    100 xp
    One-to-many merge with multiple tables    100 xp 
    

##  Merging Tables With Different Join Types
0%

Take your knowledge of joins to the next level. In this chapter, you’ll work with TMDb movie data as you learn about left, right, and outer joins. You’ll also discover how to merge a table to itself and merge on a DataFrame index.

    Left join    50 xp
    Counting missing rows with left join    100 xp
    Enriching a dataset    100 xp
    How many rows with a left join?    50 xp
    Other joins    50 xp
    Right join to find unique movies    100 xp
    Popular genres with right join    100 xp
    Using outer join to select actors    100 xp
    Merging a table to itself    50 xp
    Self join    100 xp
    How does pandas handle self joins?    50 xp
    Merging on indexes    50 xp
    Index merge for movie ratings    100 xp
    Do sequels earn more?    100 xp


##  Advanced Merging and Concatenating
0%

In this chapter, you’ll leverage powerful filtering techniques, including semi-joins and anti-joins. You’ll also learn how to glue DataFrames by vertically combining and using the pandas.concat function to create new datasets. Finally, because data is rarely clean, you’ll also learn how to validate your newly combined data structures.

    Filtering joins    50 xp
    Steps of a semi-join    100 xp
    Performing an anti-join    100 xp
    Performing a semi-join    100 xp
    Concatenate DataFrames together vertically    50 xp
    Concatenation basics    100 xp
    Concatenating with keys    100 xp
    Using the append method    100 xp
    Verifying integrity    50 xp
    Validating a merge    50 xp
    Concatenate and merge to find common songs    100 xp 
    

##  Merging Ordered and Time-Series Data
0%

In this final chapter, you’ll step up a gear and learn to apply pandas' specialized methods for merging time-series and ordered data together with real-world financial and economic data from the city of Chicago. You’ll also learn how to query resulting tables using a SQL-style format, and unpivot data using the melt method.

    Using merge_ordered()    50 xp
    Correlation between GDP and S&P500    100 xp
    Phillips curve using merge_ordered()    100 xp
    merge_ordered() caution, multiple columns    100 xp
    Using merge_asof()    50 xp
    Using merge_asof() to study stocks    100 xp
    Using merge_asof() to create dataset    100 xp
    merge_asof() and merge_ordered() differences    100 xp
    Selecting data with .query()    50 xp
    Explore financials with .query()    50 xp
    Subsetting rows with .query()    100 xp
    Reshaping data with .melt()    50 xp
    Select the right .melt() arguments    50 xp
    Using .melt() to reshape government data    100 xp
    Using .melt() for stocks vs bond performance    100 xp
    Course wrap-up    50 xp


## Inner join



**The Pandas package is a powerful tool for manipulating and transforming a in Python.  However, when working on an analysis, the data needed could be in multiple tables.  This course will focus on the vital skill of merging tables together.  




# As we start, two quick clarifications.  

First, through other courses on DataCamp, you may have learned how to import tabular data as DataFrame.  In this course, you may hear the words table and DataFrame, but they are equivalent here.  

Second, we will refer to combingning different tables together as merging tales, but note that some refer to this same process as joining.  




To help us learn about merging tables, we will use data from the city of Chicago data portal.  The city of Chicago is divided into fifty local neighborhoods called wards.  We have a table with data about the local goverment offices in each ward.  

In this example, we want to merge the local government data with census data about the population of each ward.  If we look at the wards table, we have information about the local goverment of each ward, such as the goverment office address.  The census table contains the population of each ward in 2000 and 2010, and that change as percentage.  Additionally, it includes the address for the center of each ward.  

# *******************************************************************************************************************
The two tables are related by their ward column.  We can merge them together, matching the ward number from each row of the ward table to the ward number from the census table.  

**The Pandas package has an excellent DataFrame method for performing this type of merge called .merge() method.  The .merge() method takes the first DF and merges it with the second DF.  We use on= argument to tell it we want to merge the 2 DFs on the ward column.  Since we listed the ward table first, its columns will appear first in the output, followed by the columns from the second census table.  By default, it returns the rows have matching values for the on=column in both tables.  This is called an inner join.  

An inner join will only return rows that have matching values in both tables.  

**You may have noticed that the merged table has columns with suffixes of underscore x or y.  This is because both wards and census tables contained address and zip columns.  To avoid multiple columns with the same name, they are automatically given a suffix by the merge method.  We can use the suffix= argument of the .merge() to control this behavior.  We provide a tuple where all of the overlapping columns in the left table are given the suffix '_ward', and those of the right table will be given the suffix '_cen'.  This makes it easier for us to tell the difference between the columns.   



wards = pd.read_csv('Ward_Offices.csv')
print(wards.head())
print(wards.shape)

--------------------------------------------------------------------
  ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

(50, 4)


census = pd.read_csv('Ward_Census.csv')
print(census.head())
print(census.shape)

------------------------------------------------------------------------------------
  ward  pop_2000  pop_2010 change                                  address    zip  
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET  60647  
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500  60622  
2    3     40385     53039    31%                      17 EAST 38TH STREET  60653  
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL  60653  
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE  60637

(50,6)


In [6]:
import pandas as pd



ward = pd.read_pickle('ward.p')
census = pd.read_pickle('census.p')


print(ward.head())
print(ward.shape)

print(census.head())
print(census.shape)



merged_df = ward.merge(census, how='inner', on='ward')
#print(help(pd.merge))
print(merged_df.head())
print(len(merged_df))


merged_df2 = ward.merge(census, how='inner', on='ward', suffixes=('_ward', '_cen'))
print(merged_df2.head())

  ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649
(50, 4)
  ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637  
(50, 6)
  ward            alderman                     

## What column to merge on?

Chicago provides a list of taxicab owners and vehicles licensed to operate within the city, for public safety. Your goal is to merge two tables together. One table is called taxi_owners, with info about the taxi cab company owners, and one is called taxi_veh, with info about each taxi cab vehicle. Both the taxi_owners and taxi_veh tables have been loaded for you and you can explore them in the IPython shell.

Choose the column you would use to merge the two tables on using the .merge() method.
Instructions
50 XP
Possible Answers

#    on='rid'
    on='vid'
    on='year'
    on='zip'
    

In [ ]:
In [1]:
print(taxi_owners.head())
     rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618
In [2]:
print(taxi_veh.head())
    vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO

## Your first inner join

# *******************************************************************************************************************
# You have been tasked with figuring out what the most popular types of fuel used in Chicago taxis are. 
To complete the analysis, you need to merge the taxi_owners and taxi_veh tables together on the vid column. You can then use the merged table along with the .value_counts() method to find the most common fuel_type.

Since you'll be working with pandas throughout the course, the package will be preloaded for you as pd in each exercise in this course. Also the taxi_owners and taxi_veh DataFrames are loaded for you.
Instructions 1/3
100 XP

    Question 1
    Merge taxi_owners with taxi_veh on the column vid, and save the result to taxi_own_veh.
    
    
    
    Question 2
    Set the left and right table suffixes for overlapping columns of the merge to _own and _veh, respectively.
    
    
    
    Question 3
#    Select the fuel_type column from taxi_own_veh and print the value_counts() to find the most popular fuel_types used.


In [10]:
import pandas as pd


taxi_veh = pd.read_pickle('taxi_vehicles.p')
taxi_owners = pd.read_pickle('taxi_owners.p')


print(taxi_veh.head())
print(taxi_owners.head())


taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', how='inner', suffixes=('_own', '_veh'))
print(taxi_own_veh.head())




taxi_own_veh.pivot_table(values='vid', index='fuel_type', aggfunc=lambda x: len(x.unique()))


    vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO
     rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618
     rid   vid       owner_own                 address    zip    make   model  \
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630  NISSAN  ALTIMA   
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659   HONDA     CRV   
2  T1495  1495   FU

vid
fuel_type                   
COMPRESSED NATURAL GAS    27
FLEX FUEL                 89
GASOLINE                 611
HYBRID                  2792

In [11]:
hyb_group = taxi_own_veh[taxi_own_veh['fuel_type']=='HYBRID'].copy()

print(len(hyb_group))

2792


In [13]:
# Merge the taxi_owners and taxi_veh tables setting a suffix
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', suffixes=('_own','_veh'))



# Print the value_counts to find the most popular fuel_type
print(taxi_own_veh['fuel_type'].value_counts())


# ***************************************************************************************************************** #
# This approach is much simple than my code above with pivot_table and lambda expression




HYBRID                    2792
GASOLINE                   611
FLEX FUEL                   89
COMPRESSED NATURAL GAS      27
Name: fuel_type, dtype: int64


# *******************************************************************************************************************

To count the dogs of each breed, we'll subset the breed column and use the .value_counts() method.  We can also use the sort= argument to get the breeds with biggest counts on top.  Also the normalize= argument can be used to turn the counts into proportions of the total.


# We learned that in DataCamp Data Manipulation with Pandas course


## Inner joins and number of rows returned

All of the merges you have studied to this point are called inner joins.  It is necessary to understand that inner joins only return the rows with matching values in both tables.  You will explore this further by reviewing the merge between the wards and census tables, then comparing it to merges of copies of these tables that are slightly altered, named wards_altered, and census_altered.  The first row of the wards column has been changed in the altered tables.  You will examine how this affects the merge between them. The tables have been loaded for you.

For this exercise, it is important to know that the wards and census tables start with 50 rows.
Instructions 1/3
35 XP

    Question 1
    Merge wards and census on the ward column and save the result to wards_census.
    
    
    
    Question 2
    Merge the wards_altered and census tables on the ward column, and notice the difference in returned rows.
    
    
    
    Question 3
    Merge the wards and census_altered tables on the ward column, and notice the difference in returned rows.


In [15]:
print(ward.head())
print(census.head())



wards_census = ward.merge(census, on='ward')

  ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649
  ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637  


In [ ]:
wards_altered

   ward                   alderman                            address    zip
0    61         Proco "Joe" Moreno          2058 NORTH WESTERN AVENUE  60647
1     2              Brian Hopkins         1400 NORTH  ASHLAND AVENUE  60622
2     3                 Pat Dowell            5046 SOUTH STATE STREET  60609
3     4           William D. Burns    435 EAST 35TH STREET, 1ST FLOOR  60616
4     5         Leslie A. Hairston              2325 EAST 71ST STREET  60649
5     6         Roderick T. Sawyer   8001 S. MARTIN LUTHER KING DRIVE  60619
6     7        Gregory I. Mitchell              2249 EAST 95TH STREET  60617
7     8         Michelle A. Harris    8539 SOUTH COTTAGE GROVE AVENUE  60619
8     9           Anthony A. Beale                34 EAST 112TH PLACE  60628
9    10      Susan Sadlowski Garza           10500 SOUTH EWING AVENUE  60617
10   11     Patrick Daley Thompson          3659 SOUTH HALSTED STREET  60609
11   12            George Cardenas           3476 SOUTH ARCHER AVENUE  60608
12   13                Marty Quinn            6500 SOUTH PULASKI ROAD  60629
13   14            Edward M. Burke              2650 WEST 51ST STREET  60632
14   15           Raymond A. Lopez              1650 WEST 63RD STREET  60636
15   16            Toni L. Foulkes              3045 WEST 63RD STREET  60629
16   17             David H. Moore          7313 SOUTH ASHLAND AVENUE  60636
17   18          Derrick G. Curtis            8359 SOUTH PULASKI ROAD  60652
18   19          Matthew J. O'Shea         10400 SOUTH WESTERN AVENUE  60643
19   20          Willie B. Cochran    6357 SOUTH COTTAGE GROVE AVENUE  60637
20   21    Howard B. Brookins, Jr.  9011 SOUTH ASHLAND AVENUE, UNIT B  60620
21   22              Ricardo Munoz        2500 SOUTH ST. LOUIS AVENUE  60623
22   23        Michael R. Zalewski           6247 SOUTH ARCHER AVENUE  60638
23   24         Michael Scott, Jr.           1158 SOUTH KEELER AVENUE  60624
24   25       Daniel "Danny" Solis      1800 SOUTH BLUE ISLAND AVENUE  60608
25   26          Roberto Maldonado          2511 WEST DIVISION STREET  60622
26   27        Walter Burnett, Jr.             4 NORTH WESTERN AVENUE  60612
27   28             Jason C. Ervin              2602 WEST 16TH STREET  60612
28   29           Chris Taliaferro             6272 WEST NORTH AVENUE  60639
29   30          Ariel E. Reyboras        3559 NORTH MILWAUKEE AVENUE  60641
30   31  Milagros "Milly" Santiago            2521 NORTH PULASKI ROAD  60639
31   32           Scott Waguespack         2657 NORTH CLYBOURN AVENUE  60614
32   33               Deborah Mell         3001 WEST IRVING PARK ROAD  60618
33   34           Carrie M. Austin              507 WEST 111TH STREET  60628
34   35        Carlos Ramirez-Rosa           2710 NORTH SAWYER AVENUE  60647
35   36           Gilbert Villegas                 6934 WEST DIVERSEY  60607
36   37              Emma M. Mitts           4924 WEST CHICAGO AVENUE  60651
37   38           Nicholas Sposato          3821  NORTH HARLEM AVENUE  60634
38   39           Margaret Laurino          4404 WEST LAWRENCE AVENUE  60630
39   40        Patrick J. O'Connor          5850 NORTH LINCOLN AVENUE  60659
40   41      Anthony V. Napolitano           7442 NORTH HARLEM AVENUE  60631
41   42             Brendan Reilly   325 WEST HURON STREET, SUITE 510  60654
42   43             Michelle Smith          2523 NORTH HALSTED STREET  60614
43   44                 Tom Tunney        3223 NORTH SHEFFIELD AVENUE  60657
44   45              John S. Arena        4754 NORTH MILWAUKEE AVENUE  60630
45   46            James Cappleman         4544 NORTH BROADWAY AVENUE  60640
46   47                Ameya Pawar          4243 NORTH LINCOLN AVENUE  60618
47   48             Harry Osterman         5533 NORTH BROADWAY AVENUE  60640
48   49                  Joe Moore        7356 NORTH GREENVIEW AVENUE  60626
49   50       Debra L. Silverstein    2949 WEST DEVON AVENUE, SUITE A  60659





census_altered

ward  pop_2000  pop_2010 change                                            address    zip
None     52951     56149     6%                        2765 WEST SAINT MARY STREET  60647
   2     54361     55805     3%                           WM WASTE MANAGEMENT 1500  60622
   3     40385     53039    31%                                17 EAST 38TH STREET  60653
   4     51953     54589     5%            31ST ST HARBOR BUILDING LAKEFRONT TRAIL  60653
   5     55302     51455    -7%            JACKSON PARK LAGOON SOUTH CORNELL DRIVE  60637
   6     54989     52341    -5%                               150 WEST 74TH STREET  60636
   7     54593     51581    -6%                          8549 SOUTH OGLESBY AVENUE  60617
   8     54039     51687    -4%                         1346-1352 EAST 75TH STREET  60649
   9     52008     51519    -1%                 11039-11059 SOUTH WENTWORTH AVENUE  60628
  10     56613     51535    -9%                               10534 SOUTH AVENUE F  46394
  11     64228     51497   -20%                            943-947 WEST 14TH PLACE  60607
  12     68922     52235   -24%                         CP 46 STEVENSON EXPRESSWAY  60632
  13     64382     53722   -17%                    SOUTH RAMP SOUTH LARAMIE AVENUE  60638
  14     80143     54031   -33%                              4540 WEST 51ST STREET  60632
  15     56057     51501    -8%    CHICAGO FIRE DEPARTMENT ENGINE COMPANY 123 2215  60632
  16     50205     51954     3%                             6036 SOUTH WOOD STREET  60636
  17     49264     51846     5%                       7216 SOUTH WINCHESTER AVENUE  60636
  18     55043     52992    -4%                          3286 WEST COLUMBUS AVENUE  60652
  19     54546     51525    -6%                        9999 SOUTH FRANCISCO AVENUE  60805
  20     51854     52372     1%                     DAN RYAN EXPRESSWAY PARK MANOR  60621
  21     51751     51632     0%                     8852-8854 SOUTH EMERALD AVENUE  60620
  22     59734     53515   -10%                              4233 WEST 36TH STREET  60632
  23     63691     53728   -16%  CHICAGO MIDWAY INTERNATIONAL AIRPORT WEST 62ND...  60629
  24     50879     54909     8%                       1635 SOUTH CHRISTIANA AVENUE  60623
  25     55954     54539    -3%                      1632-1746 SOUTH MILLER STREET  60608
  26     56841     53516    -6%             LITTLE CUBS FIELD COMFORT STATION 1400  60622
  27     61287     52939   -14%                      2151-2153 WEST CHICAGO AVENUE  60651
  28     49423     55199    12%                        RML SPECIALTY HOSPITAL 3435  60624
  29     61949     55267   -11%                        1241 NORTH RIDGELAND AVENUE  60302
  30     72698     55560   -24%                          5118 WEST FLETCHER STREET  60641
  31     65045     53724   -17%                          2854 NORTH KEATING AVENUE  60641
  32     57204     55184    -4%                        2901 NORTH WASHTENAW AVENUE  60618
  33     63695     55598   -13%                    4041-4043 NORTH RICHMOND STREET  60625
  34     49922     51599     3%                    11544-11546 SOUTH PEORIA STREET  60827
  35     57588     55281    -4%                           3634 WEST BELMONT AVENUE  60618
  36     63376     54766   -14%                       2918 NORTH RUTHERFORD AVENUE  60634
  37     56120     51538    -8%                         4738-4748 WEST RICE STREET  60651
  38     66011     56001   -15%                    7307-7331 WEST IRVING PARK ROAD  60706
  39     64291     55882   -13%                  QUEEN OF ALL SAINTS BASILICA 6280  60646
  40     58652     55319    -6%                         5536 NORTH ARTESIAN AVENUE  60645
  41     56127     55991     0%                          1652 SOUTH CLIFTON AVENUE  60068
  42     68102     55870   -18%                          410-420 WEST GRAND AVENUE  60654
  43     57668     56170    -3%                              LINCOLN PARK ZOO 2001  60614
  44     58758     56058    -5%                         507-513 WEST ALDINE AVENUE  60657
  45     60653     55967    -8%       CONGREGATIONAL CHURCH OF JEFFERSON PARK 5320  60630
  46     56587     53784    -5%                 UPTOWN BROADWAY BUILDING 4743-4763  60640
  47     52108     55074     6%                           2153 WEST BERTEAU AVENUE  60618
  48     56246     55014    -2%                         1025 WEST HOLLYWOOD AVENUE  60660
  49     59435     54633    -8%                             1426 WEST ESTES AVENUE  60645
  50     62383     55809   -11%                       2638 WEST NORTH SHORE AVENUE  60645

In [ ]:
# Print the first few rows of the wards_altered table to view the change 
print(wards_altered[['ward']].head())

# Merge the wards_altered and census tables on the ward column
wards_altered_census = wards_altered.merge(census, on='ward')

# Print the shape of wards_altered_census
print('wards_altered_census table shape:', wards_altered_census.shape)




# Print the first few rows of the census_altered table to view the change 
print(census_altered[['ward']].head())

# Merge the wards and census_altered tables on the ward column
wards_census_altered = wards.merge(census_altered, on='ward')

# Print the shape of wards_census_altered
print('wards_census_altered table shape:', wards_census_altered.shape)

## One-to-many relationships




**In the last lesson, we learned how to merge two DFs together with .merge() method.  In this lesson, we'll discuss different types of relationships between tables.  In particular, we'll discuss the one-to-many relationship.  But first, lets quickly consider what a one-to-one relationshipis.  


# *******************************************************************************************************************
In a one-to-one relationship, every row in the left table is related to one and only one row in the right table.  Recall the relationship between the wards table and census table we learned earlier.  

And what is a one-to-many relationship?  In a one-to-many relationship, every row in the left table is related to one or more rows in the right table.  

Within each ward, there are many businesses.  Image we will merge the wards table with a table of licensed businesses in each ward.  The business license data is stored at another table called licenses.  It holds info such as the business address and ward the business is located in.  The two DFs are related to each other by their ward column.  


# *******************************************************************************************************************
When we merge the two tables together with the .merge() method, setting the on= attribute to the column 'ward', the resulting table has both local ward data and business license data.  Notice that ward 1 and its alderman Joe is repeated in the resulting table because the licenses table has many businesses in the 1st ward.  The Pandas takes care of the one-to-many relationship for us and doesnt require anything on our end.  We can use the same syntax as we did with one-to-one relationship.  By printing the shape, we can see our original tablehas 50 rows, while after merging with the licenses table, the resulting table has 1000 rows.  When you merge tables that have a one-to-many relationship, the number of rows returned will likely be different than the number in the left table.  



In [18]:
licenses = pd.read_pickle('licenses.p')

print(licenses.head())
print(ward.head())



ward_license = ward.merge(licenses, on='ward', suffixes=('_ward', '_lic'))
print(ward_license.head())

  account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613
  ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649
  ward            alderman               address_ward zip_ward account  aid  \
0    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647 

## One-to-many classification

Understanding the difference between a one-to-one and one-to-many relationship is a useful skill. In this exercise, consider a set of tables from an e-commerce website. The hypothetical tables are the following:

    A customer table with information about each customer
    A cust_tax_info table with customers unique tax IDs
    An orders table with information about each order
    A products table with details about each unique product sold
    An inventory table with information on how much total inventory is available to sell for each product

Instructions
100XP

    Select the relationship type that is most appropriate for the relationship between the different tables: One-to-one, or One-to-many.


In [ ]:
# The relationship between products and inventory are one-to-one relationship

## One-to-many merge

A business may have one or multiple owners. In this exercise, you will continue to gain experience with one-to-many merges by merging a table of business owners, called biz_owners, to the licenses table. Recall from the video lesson, with a one-to-many relationship, a row in the left table may be repeated if it is related to multiple rows in the right table. In this lesson, you will explore this further by finding out what is the most common business owner title. (i.e., secretary, CEO, or vice president)

The licenses and biz_owners DataFrames are loaded for you.
Instructions
100 XP

    Starting with the licenses table on the left, merge it to the biz_owners table on the column account, and save the results to a variable named licenses_owners.
#    Group licenses_owners by title and count the number of accounts for each title. Save the result as counted_df
    Sort counted_df by the number of accounts in descending order, and save this as a variable named sorted_df.
    Use the .head() method to print the first few rows of the sorted_df.


In [40]:

biz_owners = pd.read_pickle('business_owners.p')


licenses_owners = licenses.merge(biz_owners, on='account')
print(licenses_owners.head())
print(licenses_owners.shape)



# ***************************************************************************************************************** #
print(licenses_owners.groupby('title')['account'].value_counts())
# ***************************************************************************************************************** #
print(licenses_owners.groupby('title')['title'].value_counts())


# ***************************************************************************************************************** #
counted_df = licenses_owners.groupby('title')['account'].count()
# ***************************************************************************************************************** #
counted_df = licenses_owners.groupby('title').agg({'account':'count'})


counted_df = counted_df.sort_values('account', ascending=False)


print(counted_df)

  account ward  aid              business              address    zip  \
0  307071    3  743  REGGIE'S BAR & GRILL      2105 S STATE ST  60616   
1      10   10  829            HONEYBEERS  13200 S HOUSTON AVE  60633   
2      10   10  829            HONEYBEERS  13200 S HOUSTON AVE  60633   
3   10002   14  775           CELINA DELI    5089 S ARCHER AVE  60632   
4   10002   14  775           CELINA DELI    5089 S ARCHER AVE  60632   

  first_name last_name      title  
0     ROBERT     GLICK     MEMBER  
1      PEARL   SHERMAN  PRESIDENT  
2      PEARL   SHERMAN  SECRETARY  
3     WALTER    MROZEK    PARTNER  
4     CELINA    BYRDAK    PARTNER  
(19497, 9)
title            account
ASST. SECRETARY  16301      3
                 57770      3
                 11071      2
                 16964      2
                 1949       2
                           ..
VICE PRESIDENT   85943      1
                 85956      1
                 86083      1
                 86202      1
         

In [ ]:
# Merge the licenses and biz_owners table on account
licenses_owners = licenses.merge(biz_owners, on='account')

# Group the results by title then count the number of accounts
counted_df = licenses_owners.groupby('title').agg({'account':'count'})

# Sort the counted_df in desending order
sorted_df = counted_df.sort_values('account', ascending=False)

# Use .head() method to print the first few rows of sorted_df
print(sorted_df.head())

## Merging multiple DataFrames




**In our last lesson, we learned how to merge two tables with a one-to-many relationship using the .merge() method.  Merging data like this is a necessary skill to bring together data from different sources to answer some more complex data questions.  Sometimes we need to merge together more than just two tables to complete our analysis.  


In the previous lesson, we used two tables form the city of Chicago.  One table contained business licenses issued by the city, the other table listed info about the local neighborhoods called wards, including the local goverment official's office.  Now we also have a table of businesses that have received small business grant money from Chicago.  
# *******************************************************************************************************************
# The grants are funded by taxpayers money.  Therefore, it would be helpful to analyze how much grant money each business received and in what ward that business is located.  We then could determine if one ward's business received a disproportionately large anount of grant money.  



To pull all of this information togerther, lets first connect our grants table to our licenses table.  The two tables are related by their company name and location.  Lets pause here for a moment.  


# *******************************************************************************************************************
If we merge the two tables only using the zip column, then the 60616 zip of Reggie's Bar from the licenses table will be matched to multiple business in the grants table with the same zip.  

Our code sample prints the first few rows and some columns of the merged table.  The output of the merge duplicates Reggie's Bar for each matching zip in the grats table, which is not what we want.  If instead, we merge on address only, there is a small risk that the address would repeat in different parts of the city.  Therefore, the best option is to merge the tables using the combination of both address and zip code.  (???????????????)

We merge the two DFs as shown before, except in this case, we pass a list of the column names we want to merge on the the on= argument.  This allows us to use multiple columns in the merge.  As before, the matching rows between the two DFs are returned with the columns from the grants table table listed first.  However, when we merge on two columns, in this case address and zip code, we are requiring that both the address and zip code of a row in the left table match the address and zip code of a row in the right table in order for them tobe linked to each other in the merge.  

We can now extend this example to third table.  First, we merge the grands table with the wards table on the ward column again, adding suffixes= to the repeated column names.  Note that we're using Python's backslash line continuation method to add the second merge on the next line.  Python will read this as just one line of code.  Don't forget your backslash.  

Now our output table has information about grants, business, and wards.  We can now complete our analysis.  We can now sum the grants by ward and plot the results.  Some wards have received more grants than others.  


We could continue to merge dditional tables as needed.  




print(licenses.head())

----------------------------------------------------------------------------
  account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613



grants = pd.read_csv('Small_Business_Grant_Agreements.csv')

print(grants.head())

-----------------------------------------------------------
    address          | zip   | grant     | company
0   1000 S Kostn...  | 60624 | 148914.50 | Nationwide F...
1   1000 W 35th ST   | 60609 | 100000.00 | Small Batch,...
2   1000 W Fulto...  | 60612 | 34412.50  | Fulton Marke...
3   10008 S West...  | 60643 | 12285.32  | Law Offices...
4   1002 W Argyl...  | 60640 | 28998.75  | Masala's Ind...
   
   
   
   
grants_licenses = grants.merge(licenses, on='zip')

# *******************************************************************************************************************
print(grants_licenses.loc[grants_licenses['business']=='Reggie's Bar & Grill', 
                         ['grant', 'company', 'account', 'ward', 'business']])
# *******************************************************************************************************************

-----------------------------------------------------------------------------
    grant     | company         | account   | ward  | business
0   136443.07 | Cedars Medit... | 307071    | 3     | Reggie's Bar & Grill
1   39943.15  | Darryl & Fyl... | 307071    | 3     | Reggie's Bar & Grill
2   31250.0   | Jgf Management  | 307071    | 3     | Reggie's Bar & Grill
3   143427.79 | Hyde Park An... | 307071    | 3     | Reggie's Bar & Grill
4   69500.00  | Zberry Inc      | 307071    | 3     | Reggie's Bar & Grill
    
    
    
    
grants.merge(licenses, on=['address', 'zip'])
    
    
    
    
grants_licenses_ward = grants.merge(licenses, on=['zip', 'address']) \ 
                                    .merge(wards, on='ward', suffixes=('_bus', '_ward'))


import matplotlib.pyplot as plt

grants_licenses_ward.groupby('ward')['grand'].agg('sum').plot(kind='bar', y='grant')
plt.show()


In [41]:
print(licenses.head())

  account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613


## Total riders in a month

# *******************************************************************************************************************
Your goal is to find the total number of rides provided to passengers passing through the Wilson station (station_name == 'Wilson') when riding Chicago's public transportation system on weekdays (day_type == 'Weekday') in July (month == 7). Luckily, Chicago provides this detailed data, but it is in three different tables. You will work on merging these tables together to answer the question. This data is different from the business related data you have seen so far, but all the information you need to answer the question is provided.

The cal, ridership, and stations DataFrames have been loaded for you. The relationship between the tables can be seen in the diagram below.

cal           ridership        stations
  year -        station_id ----- station_id
  month --    - year             station_name
  day ---    -- month            location
  day_type  --- day
                rides

Table diagram. The cal table relates to ridership via year, month, and day. The ridership table relates to the stations table via station_id.
Instructions 1/3
50 XP

    Question 1
    Merge the ridership and cal tables together, starting with the ridership table on the left and save the result to the variable ridership_cal. If you code takes too long to run, your merge conditions might be incorrect.
    
    
    
    Question 2
    Extend the previous merge to three tables by also merging the stations table.
    
    
    
    Question 3
    Create a variable called filter_criteria to select the appropriate rows from the merged table so that you can sum the rides column.


In [56]:
import pandas as pd


stations = pd.read_pickle('stations.p')
cal = pd.read_pickle('cta_calendar.p')
ridership = pd.read_pickle('cta_ridership.p')


print(stations.head())
print(cal.head())
print(ridership.head())


ridership_cal = ridership.merge(cal, on=['year', 'month', 'day'])
print(ridership_cal.sample(7))



ridership_cal_stations = ridership_cal.merge(stations, on='station_id')
print(ridership_cal_stations.sample(8))




filter_criteria = ridership_cal_stations.loc[(ridership_cal_stations['station_name']=='Wilson') & 
                                         (ridership_cal_stations['month']==7) & 
                                         (ridership_cal_stations['day_type']=='Weekday'), 'rides'].sum()#value_counts()


print('\n')
print(filter_criteria)



# To feel and to think hwo we get there, whats the best approach doing it.  Think

  station_id        station_name                 location
0      40010  Austin-Forest Park  (41.870851, -87.776812)
1      40020         Harlem-Lake  (41.886848, -87.803176)
2      40030        Pulaski-Lake  (41.885412, -87.725404)
3      40040        Quincy/Wells   (41.878723, -87.63374)
4      40050               Davis   (42.04771, -87.683543)
   year  month  day        day_type
0  2019      1    1  Sunday/Holiday
1  2019      1    2         Weekday
2  2019      1    3         Weekday
3  2019      1    4         Weekday
4  2019      1    5        Saturday
  station_id  year  month  day  rides
0      40010  2019      1    1    576
1      40010  2019      1    2   1457
2      40010  2019      1    3   1543
3      40010  2019      1    4   1621
4      40010  2019      1    5    719
     station_id  year  month  day  rides  day_type
654       41500  2019      3   14   2722   Weekday
2087      41660  2019      8   20  21143   Weekday
2830      40540  2019     11   11   5925   Weekday
2154

In [ ]:
# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']) \
            				.merge(stations, on='station_id')



# ***************************************************************************************************************** #

# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']) \
							.merge(stations, on='station_id')

# Create a filter to filter ridership_cal_stations
filter_criteria = ((ridership_cal_stations['month'] == ____) 
                   & (ridership_cal_stations['day_type'] == ____) 
                   & (ridership_cal_stations['station_name'] == ____))

# Use .loc and the filter to select for rides
print(ridership_cal_stations.loc[filter_criteria, 'rides'].sum())

# ***************************************************************************************************************** #


## Three table merge

To solidify the concept of a three DataFrame merge, practice another exercise. A reasonable extension of our review of Chicago business data would include looking at demographics information about the neighborhoods where the businesses are. A table with the median income by zip code has been provided to you. You will merge the licenses and wards tables with this new income-by-zip-code table called zip_demo.

The licenses, wards, and zip_demo DataFrames have been loaded for you.
Instructions
100 XP

#    Starting with the licenses table, merge to it the zip_demo table on the zip column. Then merge the resulting table to the wards table on the ward column. Save result of the three merged tables to a variable named licenses_zip_ward.
    Group the results of the three merged tables by the column alderman and find the median income.


In [58]:
licenses = pd.read_pickle('licenses.p')
wards = pd.read_pickle('ward.p')
zip_demo = pd.read_pickle('zip_demo.p')


print(licenses.head())
print(wards.head())
print(zip_demo.head())


licenses_zip_ward = licenses.merge(zip_demo, on='zip').merge(wards, on='zip')
licenses_zip_ward.groupby('alderman')['income'].median()


  account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613
  ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649
     zip  income
0  60630   70122
1  60640   50488
2  60622   87143
3  60614  100116
4  60608   41226


alderman
Ameya Pawar                   66246.0
Anthony A. Beale              38206.0
Anthony V. Napolitano         82226.0
Ariel E. Reyboras             33959.0
Brendan Reilly               104629.0
Brian Hopkins                 87143.0
Carlos Ramirez-Rosa           68223.0
Carrie M. Austin              38206.0
Chris Taliaferro              41307.0
Daniel "Danny" Solis          41226.0
David H. Moore                27573.0
Deborah Mell                  66246.0
Debra L. Silverstein          50554.0
Derrick G. Curtis             65770.0
Edward M. Burke               42335.0
Emma M. Mitts                 36283.0
George Cardenas               41226.0
Gilbert Villegas              92240.0
Gregory I. Mitchell           38417.0
Harry Osterman                50488.0
Howard B. Brookins, Jr.       33304.0
James Cappleman               50488.0
Jason C. Ervin                38756.0
Joe Moore                     39163.0
John S. Arena                 70122.0
Leslie A. Hairston            24941.0
Mar

## One-to-many merge with multiple tables

# *******************************************************************************************************************
# In this exercise, assume that you are looking to start a business in the city of Chicago. Your perfect idea is to start a company that uses goats to mow the lawn for other businesses. However, you have to choose a location in the city to put your goat farm. You need a location with a great deal of space and relatively few businesses and people around to avoid complaints about the smell. You will need to merge three tables to help you choose your location. The land_use table has info on the percentage of vacant land by city ward. The census table has population by ward, and the licenses table lists businesses by ward.

The land_use, census, and licenses tables have been loaded for you.
Instructions 1/3
35 XP

    Question 1
    Merge land_use and census on the ward column. Merge the result of this with licenses on the ward column, using the suffix _cen for the left table and _lic for the right table. Save this to the variable land_cen_lic.
    
    
    
    Question 2
#    Group land_cen_lic by ward, pop_2010 (the population in 2010), and vacant, then count the number of accounts. Save the results to pop_vac_lic.
    
    
    
    Question 3
    Sort pop_vac_lic by vacant, account, andpop_2010 in descending, ascending, and ascending order respectively. Save it as sorted_pop_vac_lic.



In [10]:
import pandas as pd

land_use = pd.read_pickle('land_use.p')
census = pd.read_pickle('census.p')
licenses = pd.read_pickle('licenses.p')



# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward').merge(licenses, on='ward', suffixes=('_cen', '_lic'))
print(land_cen_lic.head())




# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
                    .merge(licenses, on='ward', suffixes=('_cen','_lic'))

# Group by ward, pop_2010, and vacant, then count the # of accounts
pop_vac_lic = land_cen_lic.groupby(['ward', 'pop_2010', 'vacant'], 
                                   as_index=False).agg({'account':'count'}).sort_index(level=['ward', 'pop_2010', 'vacant'])
print(pop_vac_lic.head())




# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
                    .merge(licenses, on='ward', suffixes=('_cen','_lic'))

# Group by ward, pop_2010, and vacant, then count the # of accounts
pop_vac_lic = land_cen_lic.groupby(['ward','pop_2010','vacant'], 
                                   as_index=False).agg({'account':'count'})

# Sort pop_vac_lic and print the results
sorted_pop_vac_lic = pop_vac_lic.sort_values(['vacant', 'account', 'pop_2010'], 
                                             ascending=[False, True, True])

# Print the top few rows of sorted_pop_vac_lic
print(sorted_pop_vac_lic.head())

  ward  residential  commercial  industrial  vacant  other  pop_2000  \
0    1           41           9           2       2     46     52951   
1    1           41           9           2       2     46     52951   
2    1           41           9           2       2     46     52951   
3    1           41           9           2       2     46     52951   
4    1           41           9           2       2     46     52951   

   pop_2010 change                  address_cen zip_cen account  aid  \
0     56149     6%  2765 WEST SAINT MARY STREET   60647   12024  NaN   
1     56149     6%  2765 WEST SAINT MARY STREET   60647   14446  743   
2     56149     6%  2765 WEST SAINT MARY STREET   60647   14624  775   
3     56149     6%  2765 WEST SAINT MARY STREET   60647   14987  NaN   
4     56149     6%  2765 WEST SAINT MARY STREET   60647   15642  814   

               business              address_lic zip_lic  
0   DIGILOG ELECTRONICS       1038 N ASHLAND AVE   60622  
1      EMPTY BOT

# Sort index we have to specify level=[columns], not needed when we apply it to the sort_values()

## Left join




**In this lesson, we'll discuss how a left join works, which is another way to merge two tables.  Before we start about left joins, lets quickly review what we have learned so far.  

In chapter 1, we introduced the Pandas df.merge() method that allows us to combine two tables by specifying one or more key columns to link  the tables by.  By default, the .merge() method performs an inner join, returning only the rows of data with matching values in the key columns of both tables.  

# *******************************************************************************************************************
In this lesson, we'll talk about the idea of a left join.  A left join returns all rows of data from the left table and only those rows from the right table where key columns match.  


To help us learn more about left joins and other concepts in this chapter, we'll use data from The Movie Database, a company-built movie database  with info on thousands of movies, their casts, and popularity.  In our next example, we have two tables from The Movie Database that we want to merge.  

Our first table, named movies, holds info about individual movies such as the title name and its popularity.  Additionally, each movie is given an ID number.  Our second table is named taglines, which contains a movie ID number and the tag line for the movie.  To merge these tables with a left join, we use our merge method similar to what we learned in chapter 1.  Here we list the movie table first and merge  it to the taglines table on the on='id' column in both tables.  However, notice an additional argument named how=, this argument defines how to merge the two tables.  Remember that Pandas use NaN to denote missing data.  

After the merge, our resulting table has a 4805 rows.  This is because we are returning all of the rows of data from the movies table, and the relationship between the movies table and taglines table is one-to-one.  Therefore, in a one-to-one merge like this, a left join will always return the same number rows as the left table.  




left table                     right table                      result table

------------------------      -------------------              ----------------------------------
 A      | B      | C          C       | D                       A      | B     | C     | D
 A2     | B2     | C2         C1      | D1                      A2     | B2    | C2    | D1
 A3     | B3     | C3         C2      | D2                      A3     | B3    | C3    |
 A4     | B4     | C4         C4      | D4                      A4     | B4    | C4    | D4
                              C5      | D5


movies = pd.read_csv('tmdb_movies.csv')
print(movies.head())
print(movies.shape)

------------------------------------------------------------
      id                 title  popularity release_date
0    257          Oliver Twist   20.415572   2005-09-23
1  14290  Better Luck Tomorrow    3.877036   2002-01-12
2  38365             Grown Ups   38.864027   2010-06-24
3   9672              Infamous    3.680896   2006-11-16
4  12819       Alpha and Omega   12.300789   2010-09-17





In [1]:
import pandas as pd

movies = pd.read_pickle('movies.p')
print(movies.head())
print(movies.shape)


taglines = pd.read_pickle('taglines.p')
print(taglines.head())
print(taglines.shape)


print('\n')
movies_taglines = movies.merge(taglines, on='id', how='left')
print(movies_taglines.head())
print(movies_taglines.shape)

      id                 title  popularity release_date
0    257          Oliver Twist   20.415572   2005-09-23
1  14290  Better Luck Tomorrow    3.877036   2002-01-12
2  38365             Grown Ups   38.864027   2010-06-24
3   9672              Infamous    3.680896   2006-11-16
4  12819       Alpha and Omega   12.300789   2010-09-17
(4803, 4)
       id                                         tagline
0   19995                     Enter the World of Pandora.
1     285  At the end of the world, the adventure begins.
2  206647                           A Plan No One Escapes
3   49026                                 The Legend Ends
4   49529            Lost in our world, found in another.
(3955, 2)


      id                 title  popularity release_date  \
0    257          Oliver Twist   20.415572   2005-09-23   
1  14290  Better Luck Tomorrow    3.877036   2002-01-12   
2  38365             Grown Ups   38.864027   2010-06-24   
3   9672              Infamous    3.680896   2006-11-16   

## Counting missing rows with left join

The Movie Database is supported by volunteers going out into the world, collecting data, and entering it into the database. This includes financial data, such as movie budget and revenue. If you wanted to know which movies are still missing data, you could use a left join to identify them. Practice using a left join by merging the movies table and the financials table.

The movies and financials tables have been loaded for you.
Instructions 1/3
35 XP

    Question 1
What column is likely the best column to merge the two tables on?
Possible Answers

    on='budget'
    on='popularity'
    on='id'
    
    
    
    Question 2     Missing
    
    
    
    
    Question 3     Missing
    
    

In [2]:

financials = pd.read_pickle('financials.p')
print(financials.head())
print(financials.shape)


print(movies.head())
print(movies.shape)


movies_financials = movies.merge(financials, on='id', how='left')
print(movies_financials.head())
print(movies_financials.shape)



print('\n')
print(movies_financials.isna().sum())


# If you wanted to know which movies are still missing data, you could use a left join to identify them.



       id     budget       revenue
0   19995  237000000  2.787965e+09
1     285  300000000  9.610000e+08
2  206647  245000000  8.806746e+08
3   49026  250000000  1.084939e+09
4   49529  260000000  2.841391e+08
(3229, 3)
      id                 title  popularity release_date
0    257          Oliver Twist   20.415572   2005-09-23
1  14290  Better Luck Tomorrow    3.877036   2002-01-12
2  38365             Grown Ups   38.864027   2010-06-24
3   9672              Infamous    3.680896   2006-11-16
4  12819       Alpha and Omega   12.300789   2010-09-17
(4803, 4)
      id                 title  popularity release_date      budget  \
0    257          Oliver Twist   20.415572   2005-09-23  50000000.0   
1  14290  Better Luck Tomorrow    3.877036   2002-01-12         NaN   
2  38365             Grown Ups   38.864027   2010-06-24  80000000.0   
3   9672              Infamous    3.680896   2006-11-16  13000000.0   
4  12819       Alpha and Omega   12.300789   2010-09-17  20000000.0   

       

## Enriching a dataset

Setting how='left' with the .merge()method is a useful technique for enriching or enhancing a dataset with additional information from a different table. In this exercise, you will start off with a sample of movie data from the movie series Toy Story. Your goal is to enrich this data by adding the marketing tag line for each movie. You will compare the results of a left join versus an inner join.

The toy_story DataFrame contains the Toy Story movies. The toy_story and taglines DataFrames have been loaded for you.
Instructions 1/2
50 XP

    Question 1
    Merge toy_story and taglines on the id column with a left join, and save the result as toystory_tag.
    
    
    
    Question 2
    With toy_story as the left table, merge to it taglines on the id column with an inner join, and save as toystory_tag.


In [11]:
print(movies.head())

print(movies[movies['title']=='Toy Story'].copy())

toy_story = movies[movies.filter(like ='Toy Story', axis=1)]
print(toy_story)


# Merge the toy_story and taglines tables with a left join
toystory_tag = toy_story.merge(taglines, on='id', how='left')

# Print the rows and shape of toystory_tag
print(toystory_tag.)
print(toystory_tag.shape)

SyntaxError: invalid syntax (3843563265.py, line 13)